### PR TITLE
8339728 : [Accessibility,Windows,JAWS] Bug in the getKeyChar method of the AccessBridge class

### PR DESCRIPTION
**Pull request by kumarabhi006**
For a JMenuItem with a shortcut like _Ctrl + Comma_, the `getKeyChar` method of the `AccessBridge` class cuts the _Comma_ text to the first character and hence transfers `C (instead of Comma)` via the `AccessBridge` API. For a shortcut _Ctrl + Comma_ in a menu item, screen readers announce _Ctrl + C_ instead of _Ctrl + Comma_ as shortcut. Same issue exists for **Enter, Period and other keys** as well where only the first character is returned from the getKeyChar method. 

Proposed fix is to ensure that the _char representation of the shortcuts_ is returned by the `getKeyChar` method and it worked fine except for **"Enter"** key. It is unclear to me why it has not announced by AT (JAWS) but I think that may be due to the char representation of the Enter key which is a Line Feed.

AT was also not able to announce the _Tab and Space_ key as a shortcut. Adding these keys in the control key list in _AccessBridge_ and in supported control code list in _AccessBridgePackages_ files enabled them to be announced by AT.

Manual test case is added to verify the shortcut for JMenuItems.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8339728](https://bugs.openjdk.org/browse/JDK-8339728): [Accessibility,Windows,JAWS] Bug in the getKeyChar method of the AccessBridge class (**Bug** - P3)

### Reviewers
 * [Alexander Zuev](https://openjdk.org/census#kizune) (@azuev-java - **Reviewer**) Review applies to [b6169d3d](https://git.openjdk.org/jdk/pull/22822/files/b6169d3d895958acd677dc11ab6a02c863052fd7)
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**)
 * [Prasanta Sadhukhan](https://openjdk.org/census#psadhukhan) (@prsadhuk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22822/head:pull/22822` \
`$ git checkout pull/22822`

Update a local copy of the PR: \
`$ git checkout pull/22822` \
`$ git pull https://git.openjdk.org/jdk.git pull/22822/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22822`

View PR using the GUI difftool: \
`$ git pr show -t 22822`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22822.diff">https://git.openjdk.org/jdk/pull/22822.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22822#issuecomment-2552814098)
</details>

### CODE REVIEW

**Comment by kumarabhi006** [comment]
@azuev-java Please review.

File: test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import javax.swing.JMenuBar;
+import javax.swing.JMenuItem;
+import javax.swing.KeyStroke;
+
+/*
+ * @test
+ * @bug 8339728
+ * @summary Tests that JAWS announce the shortcuts for JMenuItems.
+ * @library /java/awt/regtesthelpers
+ * @build PassFailJFrame
```
Comment (azuev-java) [inline test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java:37]: Since instructions mention JAWS i would either make this test Windows specific or if you want to test t on mac too i would rephrase the instructions so they are not OS specific. Either way will work for me.

File: src/jdk.accessibility/windows/classes/com/sun/java/accessibility/internal/AccessBridge.java [hunk 2]
```diff
@@ -3967,13 +3969,11 @@ private char getKeyChar(KeyStroke keyStroke) {
...
-        debugString("[INFO]:   Shortcut is: " + keyText);
-        if (keyText != null || keyText.length() > 0) {
-            CharSequence seq = keyText.subSequence(0, 1);
-            if (seq != null || seq.length() > 0) {
-                return seq.charAt(0);
-            }
+
+        keyCode = keyStroke.getKeyCode();
+        debugString("[INFO]:   Shortcut is: " + Integer.toHexString(keyCode));
+        if (keyCode != 0) {
```
Comment (azuev-java) [inline src/jdk.accessibility/windows/classes/com/sun/java/accessibility/internal/AccessBridge.java:3975]: Why do we need this check? If keyCode is not zero we return keyCode but if it is zero we will still return keyCode because we return zero after the condition.

File: test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java (hunk 1 above)
Comment (kumarabhi006) [reply test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java:38]: Updated.

File: src/jdk.accessibility/windows/classes/com/sun/java/accessibility/internal/AccessBridge.java (hunk 2 above)
Comment (kumarabhi006) [reply src/jdk.accessibility/windows/classes/com/sun/java/accessibility/internal/AccessBridge.java]: Yeah, the condition check is not needed. Updated.

**Review by azuev-java — APPROVED** [review]
This version looks good. Manually tested with both JAWS and NVDA haven't found no functional regressions.

**Comment by victordyakov** [comment]
@prsadhuk please review

File: src/jdk.accessibility/windows/classes/com/sun/java/accessibility/internal/AccessBridge.java [hunk 3]
```diff
@@ -3967,15 +3969,10 @@ private char getKeyChar(KeyStroke keyStroke) {
...
-            CharSequence seq = keyText.subSequence(0, 1);
-            if (seq != null || seq.length() > 0) {
-                return seq.charAt(0);
-            }
-        }
-        return 0;
+
+        keyCode = keyStroke.getKeyCode();
+        debugString("[INFO]:   Shortcut is: " + Integer.toHexString(keyCode));
+        return (char) keyCode;
```
Comment (aivanov-jdk) [inline src/jdk.accessibility/windows/classes/com/sun/java/accessibility/internal/AccessBridge.java:3975]:
```suggestion
        return (char)keyCode;
```
For consistency with the `return` statements above which have no space after the cast.

File: test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+        menu.add(menuItem1);
+        menu.add(menuItem2);
+        menu.add(menuItem3);
+        menu.add(menuItem4);
+        menu.add(menuItem5);
+        menu.add(menuItem6);
+        menu.add(menuItem7);
+
+        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
```
Comment (aivanov-jdk) [inline test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java:107]: You should not define a default close operation for a manual test, the `PassFailJFrame` framework handles closing the test window and fails the test.

File: test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                .testUI(TestJMenuItemShortcutAccessibility::createUI)
+                .build()
+                .awaitAndCheck();
+    }
+
+    private static JFrame createUI() {
+        JFrame frame = new JFrame("A Frame with Menu");
+
+        JMenuBar menuBar = new JMenuBar();
+        JMenu menu = new JMenu("Menu with Keystrokes");
```
Comment (aivanov-jdk) [inline test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java:65]:
```suggestion
        JMenu menu = new JMenu("Menu with shortcuts");
```
You refer to concept as “shortcuts” in the instructions.

File: test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+ */
+
+public class TestJMenuItemShortcutAccessibility {
+    public static void main(String[] args) throws Exception {
+        String INSTRUCTIONS = """
+                1. Start the JAWS aplication
+                2. Press Alt + M to open application Menu
+                3. Navigate the Menu Items by using UP / DOWN arrow key
+                4. Press Pass if you are able to hear correct JAWS announcements
+                   for each menu item shorcut else Fail.
```
Comment (aivanov-jdk) [inline test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java:49]:
```suggestion
                   for each menu item shortcut else Fail.
```
Typo.

You should probably remove the period… to be consistent. Either add ending punctuation to each instruction item, nor don't add it at all, including the last item.

**Review by aivanov-jdk — APPROVED** [review]
Looks good to me except for minor comments.

You may want to bump up the copyright year to 2025 now.

File: test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java (hunk 6 above)
Comment (kumarabhi006) [reply test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java]: Updated

File: test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java (hunk 4 above)
Comment (kumarabhi006) [reply test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java]: Updated

File: src/jdk.accessibility/windows/native/include/bridge/AccessBridgePackages.h [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2005, 2015, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2005, 2024, Oracle and/or its affiliates. All rights reserved.
```
Comment (prsadhuk) [inline src/jdk.accessibility/windows/native/include/bridge/AccessBridgePackages.h:2]: this copyright also needs to be updated..

File: test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+ * @requires os.family == "windows"
+ * @library /java/awt/regtesthelpers
+ * @build PassFailJFrame
+ * @run main/manual TestJMenuItemShortcutAccessibility
+ */
+
+public class TestJMenuItemShortcutAccessibility {
+    public static void main(String[] args) throws Exception {
+        String INSTRUCTIONS = """
+                1. Start the JAWS aplication
```
Comment (prsadhuk) [inline test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java:45]: "application"

File: test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+                2. Press Alt + M to open application Menu
+                3. Navigate the Menu Items by using UP / DOWN arrow key
+                4. Press Pass if you are able to hear correct JAWS announcements
+                   for each menu item shortcut else Fail
```
Comment (prsadhuk) [inline test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java:49]: should we need not expand what is "correct" JAWS announcements? like mentioning it should read aloud "full" shortcut text and not only the 1st character of shortcut

File: src/jdk.accessibility/windows/native/include/bridge/AccessBridgePackages.h (hunk 7 above)
Comment (kumarabhi006) [reply src/jdk.accessibility/windows/native/include/bridge/AccessBridgePackages.h]: Updated.

File: test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java (hunk 9 above)
Comment (kumarabhi006) [reply test/jdk/javax/accessibility/TestJMenuItemShortcutAccessibility.java]: Updated test instruction.

**Review by prsadhuk — APPROVED** [review]

**Review by aivanov-jdk — APPROVED** [review]

**Comment by aivanov-jdk** [comment]
I ran the test with Java 21 and with the fix, and I confirm JAWS announces the shortcuts correctly now.

> …it worked fine except for <kbd>Enter</kbd> key…

Should we submit a bug? It's _a known problem_ now, but there's no fix for it. It could be a limitation that may be worth calling out in a release note.

**Comment by kumarabhi006** [comment]
> > …it worked fine except for Enter key…
> 
> Should we submit a bug? It's _a known problem_ now, but there's no fix for it. It could be a limitation that may be worth calling out in a release note.

Yeah, as of now I don't have a fix for "Enter" key shortcut or it may be a limitation.
I am not sure if filing a bug is a good idea as it may just increase the backlog with no potential solution.
